### ISSUE TITLE
Support http_auth object for aiohttp connection (to support AWS Sigv4)

### ISSUE THREAD

**Issue by HoneyryderChuck** [elastic/elasticsearch-py#1584]
TL;DR: support `http_auth` as an object for async-es.

Using Elasticsearch in AWS requires the usage of AWS Sigv4 for authenticating requests. [AWS documents how to do this](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-request-signing.html#es-request-signing-python). However, the documentation is only valid for "sync" python, as the `http_auth` parameter, when used in "async" mode, only works if `http_auth` is either a string or a tuple that can be used for basic auth.

In order for [reusing the auth middleware which piggybacks on requests request lifecycle](https://github.com/tedder/requests-aws4auth/blob/master/requests_aws4auth/aws4auth.py#L318), we'd have to pass it a fully formed request. This can be done in this lib, if the aiohttp session is created [with a custom request class](https://github.com/aio-libs/aiohttp/blob/1a4126a2c0e52b331aa953288e844bac680290fc/aiohttp/client.py#L203) which "hooks the `#send` function with the auth module.

**Comment by sethmlarson** [comment]
The Elasticsearch clients go through rigorous testing and validation against the official Elasticsearch distribution coming from Elastic. Since the AWS Elasticsearch Service has incompatible APIs with Elasticsearch itself, either by missing APIs or has incompatible format, we do not support it. Please consider using [Elastic Cloud](https://cloud.elastic.co/) for an official and formally supported hosted Elasticsearch distribution.
